### PR TITLE
Add countdown barrier

### PR DESCRIPTION
**Pull request by prolic**
references https://github.com/amphp/amp/pull/246 and https://github.com/amphp/amp/issues/244

### CODE REVIEW

File: src/CountdownBarrier.php [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        return $this->currentCount;
+    }
+
+    public function getInitialCount(): int
+    {
+        return $this->initialCount;
+    }
+
+    public function signal(int $signalCount = 1): bool
```
Comment (enumag) [inline src/CountdownBarrier.php:50]: In my opinion this should return void. If you need to check that the signal resolved the promise, you can call `getCurrentCount` right after.

File: src/CountdownBarrier.php [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * @param int $signalCount
+     *
+     * @return void
```
Comment (enumag) [inline src/CountdownBarrier.php:78]: This package requires php 7.1, you can use void return type.

File: src/CountdownBarrier.php [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+     *
+     * @throws Error
+     */
+    public function reset(int $signalCount = null)
```
Comment (enumag) [inline src/CountdownBarrier.php:102]:
```suggestion
    public function reset(?int $signalCount = null)
```

File: src/CountdownBarrier.php [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            $this->deferred->resolve(true);
+
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * @param int $signalCount
```
Comment (enumag) [inline src/CountdownBarrier.php:76]: useless annotation

File: test/CountdownBarrierTest.php [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+namespace Amp\Sync\Test;
+
+use Amp\Loop;
+use Amp\Sync\CountdownBarrier;
+use Error;
+use PHPUnit\Framework\TestCase;
+
+class CountdownBarrierTest extends TestCase
+{
+    public function testSignalUntilResolved()
```
Comment (enumag) [inline test/CountdownBarrierTest.php:12]: `: void` ?

**Comment by prolic** [comment]
thanks @enumag, changes are applied

File: src/CountdownBarrier.php [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        }
+
+        if ($signalCount > $this->currentCount) {
+            throw new Error('Signal count cannot be greater than current count');
+        }
+
+        $this->currentCount -= $signalCount;
+
+        if (0 === $this->currentCount) {
+            $this->deferred->resolve(true);
```
Comment (enumag) [inline src/CountdownBarrier.php:68]: I don't see any point in resolving with true. The promise should be Promise<void> I think.

File: src/CountdownBarrier.php [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+        if (0 === $this->currentCount) {
+            throw new Error('CountdownBarrier already resolved');
+        }
+
+        $this->currentCount += $signalCount;
+    }
+
+    /** @throws Error */
+    public function reset(?int $signalCount = null): void
```
Comment (enumag) [inline src/CountdownBarrier.php:87]: I'm considering whether calling this should maybe fail the original promise if it wasn't resolved yet. Thoughts?

File: test/CountdownBarrierTest.php [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+use Amp\Loop;
+use Amp\Sync\CountdownBarrier;
+use Error;
+use PHPUnit\Framework\TestCase;
+
+class CountdownBarrierTest extends TestCase
+{
+    public function testSignalUntilResolved(): void
+    {
+        Loop::run(function () {
```
Comment (trowski) [inline test/CountdownBarrierTest.php:14]: You can use `AsyncTestCase` in this package to remove the need to call `Loop::run()` in tests.

File: src/CountdownBarrier.php (hunk 7 above)
Comment (prolic) [reply src/CountdownBarrier.php]: agree, removed

File: test/CountdownBarrierTest.php (hunk 8 above)
Comment (prolic) [reply test/CountdownBarrierTest.php]: done

File: src/CountdownBarrier.php (hunk 6 above)
Comment (prolic) [reply src/CountdownBarrier.php]: done

**Comment by prolic** [comment]
all requested changes applied and rebased

File: src/CountdownBarrier.php [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        }
+
+        $this->initialCount = $initialCount;
+        $this->currentCount = $initialCount;
+        $this->deferred = new Deferred();
+    }
+
+    public function getCurrentCount(): int
+    {
+        return $this->currentCount;
```
Comment (kelunik) [inline src/CountdownBarrier.php:42]: Should we name it getRemainingCount to make it even clearer that it counts down?

Comment (prolic) [reply src/CountdownBarrier.php:42]: done

File: src/CountdownBarrier.php [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        return $this->remainingCount;
+    }
+
+    public function getInitialCount(): int
+    {
+        return $this->initialCount;
+    }
+
+    /** @throws Error */
+    public function signal(int $signalCount = 1): void
```
Comment (kelunik) [inline src/CountdownBarrier.php:51]: `arrive` would be another possible name.

File: src/CountdownBarrier.php [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    {
+        if ($signalCount < 1) {
+            throw new Error('Signal count must be greater or equals 1');
+        }
+
+        if (0 === $this->remainingCount) {
+            throw new Error('CountdownBarrier already resolved');
+        }
```
Comment (kelunik) [inline src/CountdownBarrier.php:59]: This `if` can be removed due to the `if` below, no?

File: src/CountdownBarrier.php [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+
+        if ($signalCount > $this->remainingCount) {
+            throw new Error('Signal count cannot be greater than current count');
+        }
+
+        $this->remainingCount -= $signalCount;
+
+        if (0 === $this->remainingCount) {
```
Comment (kelunik) [inline src/CountdownBarrier.php:67]:
```suggestion
        if ($this->remainingCount === 0) {
```

File: src/CountdownBarrier.php (hunk 10 above)
Comment (kelunik) [reply src/CountdownBarrier.php:51]: Should `signal` return the promise as well?

**Comment by kelunik** [comment]
Thanks! I'll going to change the API a bit after merging now.

**Pull request merged**